Add line groups to navitia

eturck opened the pull request:
This Pull Request add the concept of line_groups in navitia based on "Group of lines" in transmodel.
Things added : 
- 2 new tables in ed : 
  - line_group (id, uri, name),
  - line_group_link (group_id, line_id, is_main_line).
- 2 more files imported in fusio2ed : 
  - line_groups.txt (line_group_id, line_group_name, main_line_id),
  - line_group_links.txt (line_group_id, line_id).
- Necessary code to write to ed, read from ed and load in kraken,
- Necessary code to write to protobuf in kraken and read protobuf in jormungandr,
- New URL in jormungandr : http://<ip>/coverage/<instance>/v1/line_groups
  ![line_groups](https://cloud.githubusercontent.com/assets/5136553/8086759/4ab1367c-0f97-11e5-8096-b382e9997172.png)
- Output group in v1/lines if the line is part of one or more 
  ![line_with_group](https://cloud.githubusercontent.com/assets/5136553/8086781/739ba950-0f97-11e5-876a-5545ee6be12b.png)
- Unit test for fusio2ed, jormungandr integration and ptref.

Some .proto files have been modified but they are stored in another repository. It's necessary to merge the pull request from this repository (navitia-proto) first, since in this one, source/navitia-proto refers to a new commit. ( see https://github.com/CanalTP/navitia-proto/pull/1 )

Waiting for your feedback, let me know if there is anything to change / improve :) !

Commit:
ed: Add alembic files for line_groups

Commit:
ed: import line_groups from NTFS files

Commit:
ed: Read line_groups from ed to nav binary file

Bump version number to 39 since serialized data are modified

Commit:
kraken: export line_groups to protobuf

Commit:
jormungandr: read line_groups from protobuf and expose line_groups API

Commit:
ed: unit test for line_groups

Commit:
ptreferential: add line_groups in the ptref_graph

... and fix ptref_test

Commit:
jormungandr: add integration tests for line_groups

buildbotctp commented:
Can one of the admins verify this patch?

eturck commented:
It's failing because of navitia-proto.
I made the PR : https://github.com/CanalTP/navitia-proto/pull/1

antoine-de commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -657,6 +657,76 @@ void LineFusioHandler::handle_line(Data& data, const csv_row& row, bool is_first
+        return;
+    }
+
+    if(line_group->second->linkedLinesURI.find(line->second->uri) != line_group->second->linkedLinesURI.end()) {
```
not sure, but it seems that the linkedLinesURI are only used to make sure no duplicates are added.

For this I prefer to add a std::map<line, std::set<>> in the class GtfsData. It's clearer that way, what is only need during data import and what is part on the object model.

antoine-de commented:
Nice PR lot's of added test :+1: !

Sorry, I haven't the time to finish the reading, I'll continue on monday.

Thanks!

Commit:
ed: move data to avoid duplicate links in gtfs_data

Remove linkedLines set in data model since it was used only during the import process.
+ Wording on fusioparset_test

eturck commented:
Thank you! It was indeed there only to prevent duplicate entries for line_group_links.

I have a generic question on the pull request workflow. Since my original commits 2 days ago, a few commits have been made on the dev branch, causing conflicts with my code.
What should I do : 
- Rebase + push force on my branch to keep the code up to date ?
- Merge dev into my branch (history will not be super clean I guess if I do that) ?
- Wait until the end of the PR to do one those ?
- Something else I didn't think of / don't know about :) ?

Have a nice day / week end !

antoine-de commented:
yes you should rebase + force push on your fork, this way the git history will be cleaner and since the force push is only on you repository, it's not possible to break the central repository.

You can either to this now or at the end of the PR, it's up to you, but I think it's easier not to wait too long not to have too many conflicts.

Commit:
jormungandr: change line_groups output to be consistent with other services

Remove GroupLink / LineLink concepts
Update navitia-proto after rebase

stifoon commented:
i would like to interact with the json interface of a "http://coverage/v1/line_groups" request.
The "lines" property does not contain lines, but a couple [line, "is_main_line"]. So it is not consistent with the other services.
Perharps a json like this one should be more consistent:

```
{
  "line_groups": [
    {
      "id": "line_group:bob",
      "name": "Direction Bob",
      "lines": [
        {
          "code": "B",
          "name": "Blue Bob",
          "color": "Blue",
          "geojson": "",
          "id": "line:bob:blue"
        },
        {
          "code": "R",
          "name": "Red Bob",
          "color": "Red",
          "geojson": "",
          "id": "line:bob:red"
        }
      ],
      "main_line": {
        "type": "line",
        "id": "line:bob:blue"
      }
    },
    {
      "id": "line_group:bobette",
      "name": "Direction Bobette",
      "lines": [
        {
          "code": "Y",
          "name": "Yellow Bobette",
          "color": "AAFFBB",
          "geojson": "",
          "id": "line:bobette:yellow"
        },
        {
          "code": "G",
          "name": "Green Bobette",
          "color": "AAFFBB",
          "geojson": "",
          "id": "line:bobette:green"
        }
      ],
      "main_line": {
        "type": "line",
        "id": "line:bobette:yellow"
      }
    }
  ]
}
```

eturck commented:
Thank you for the feedback.
It seems to be a better way to display line_groups, I'm going to make the necessary changes.

When I display groups in a "v1/lines" request, with a depth of 1, I'm going to change the current output to : 

``` json
lines: [
   {
       "id": "line:1",
        "groups": [
            {
               "id": "group:1",
               "name": "1st group",
               "main_line": {
                  "id": "line:1"
               }
            }
        ]
   }
]
```

Is this ok ? Or would you prefer to have the "main_line" only if we have a depth > 1 ?

eturck commented:
I've updated the code here and in the navitia-proto PR.

Here is the new output for line_groups : 
![line_groups](https://cloud.githubusercontent.com/assets/5136553/8227458/ecd5e040-15a7-11e5-8b99-7533264db306.png)

And for lines (I decided to not display the main_line here, it seemed better that way and the depth parameter is here if we want it) : 
![line_with_group](https://cloud.githubusercontent.com/assets/5136553/8227465/faa92b0a-15a7-11e5-8c9b-991ac54b6b10.png)
With depth=2 groups are displayed with its lines and main_line.

I'm working on providing you a link to test the json output live, but as you may know we had a few problem building on debian 7.8 the past few days... I have fallen a bit behind on that.

TeXitoi commented on fixtures/ed/ntfs/line_groups.txt:
```diff
@@ -0,0 +1,4 @@
+line_group_id,line_group_name,main_line_id
+lg1,"Groupe lignes B",l2
+lg2,"Groupe ligne Bs seule",l3
+lg3,"Groupe inconnu",iddeligneprincipalequinexistepas
```
Maybe in english?

TeXitoi commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -657,6 +657,77 @@ void LineFusioHandler::handle_line(Data& data, const csv_row& row, bool is_first
+void LineGroupFusioHandler::handle_line(Data& data, const csv_row& row, bool) {
+    if(!(has_col(id_c, row) && has_col(name_c, row) && has_col(main_line_id_c, row))) {
+        LOG4CPLUS_WARN(logger, "LineGroupFusioHandler: Line ignored in " << csv.filename <<
+            " missing line_group_id, line_group_name or main_line_id column");
```
you want to return here, no?

eturck replied:
Indeed. Thanks !

antoine-de commented:
your new json seems nice!

maybe you should call the line groups of the line 'line_groups' and not 'groups' (we try to name each collection with the type of the object).

And in the line_groups api we could:
- display only id and name for depth=0 as you say
- display the lines and main lines on depth=1
- force the main_line depth to be 0 not to duplicate too many information ?

would that be ok with you @stifoon  ?

antoine-de commented on source/ed/connectors/gtfs_parser.h:
```diff
@@ -106,6 +107,9 @@ struct GtfsData {
     std::unordered_map<std::string, std::string> comments_id_map;
 
+    // Store lines linked for each group to avoid duplicates
+    std::unordered_map<std::string, std::vector<std::string>> linked_lines_by_line_group_uri;
```
since it's to avoid duplicate, isn't a set better than the vector ?

eturck replied:
It's only to store uris of lines of each group. I don't really mind duplicate inside the vector, I just use "find" before creating a link between the group and the line.
And since I read http://lafstern.org/matt/col1.pdf I try to not use them too much =P.

TeXitoi replied:
In this case, you should use a `boost::flat_set`, a `std::set` interface with a sorted vector behind.

http://www.boost.org/doc/libs/1_58_0/doc/html/container/non_standard_containers.html#container.non_standard_containers.flat_xxx

eturck commented:
@antoine-de on the last part, on depth=1, lines and main_line are displayed, but already with a depth of 0 (in pb_converter.cpp I use depth - 1).
If we want to reduce the output size I can maybe have for the main_line : 
- If depth=0, no output,
- If depth=1, just output id / name (it should be possible since all the field in the proto file are optional I guess ?),
- If depth > 1 use fill_pb_object with depth - 1.

antoine-de commented:
IMHO, we should display the main line for depth > 1, but not with depth=(depth -1), with depth=0, since all the information is redundant.

TeXitoi commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -657,6 +657,77 @@ void LineFusioHandler::handle_line(Data& data, const csv_row& row, bool is_first
+void LineGroupLinksFusioHandler::handle_line(Data& data, const csv_row& row, bool) {
+    if(!(has_col(line_group_id_c, row) && has_col(line_id_c, row))) {
+        LOG4CPLUS_WARN(logger, "LineGroupLinksFusioHandler: Line ignored in " << csv.filename <<
+        " missing line_group_id or line_id column");
```
same here

TeXitoi commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -657,6 +657,77 @@ void LineFusioHandler::handle_line(Data& data, const csv_row& row, bool is_first
+    }
+
+    std::vector<std::string> linked_lines = gtfs_data.linked_lines_by_line_group_uri[line_group->second->uri];
+    if(std::find(linked_lines.begin(), linked_lines.end(), line->second->uri) != linked_lines.end()) {
```
Why linked line is not a set?

antoine-de commented:
Jenkins, test this please

TeXitoi commented on source/ed/ed_persistor.cpp:
```diff
@@ -726,6 +730,35 @@ void EdPersistor::insert_lines(const std::vector<types::Line*>& lines){
+    }
+    this->lotus.finish_bulk_insert();
+
+    this->lotus.prepare_bulk_insert("navitia.line_group_link", {"group_id","line_id","is_main_line"});
```
Why is there a `is_main_line` in this relation table, and not an id in the `line_group` table with the id of the main line, as it is in the NTFS? It seems cleaner and simpler to me. (Same for `types::LineGroupLinks`)

TeXitoi commented on source/ed/ed_reader.cpp:
```diff
@@ -550,6 +551,35 @@ void EdReader::fill_lines(nt::Data& data, pqxx::work& work){
+        nt::LineGroup* line_group = new nt::LineGroup();
+        const_it["uri"].to(line_group->uri);
+        const_it["name"].to(line_group->name);
+        this->line_group_map[const_it["id"].as<idx_t>()] = line_group;
```
you can also use .to() here no?

TeXitoi commented on source/ed/types.cpp:
```diff
@@ -114,6 +114,10 @@ bool Line::operator<(const Line& other) const {
 
+bool LineGroup::operator<(const LineGroup& other) const {
+    return this->name < other.name || (this->name == other.name && this < &other);
+}
```
I don't like comparing pointers. As LineGroup is a Header, is has an idx on which we can compare instead of the address (which is not the same accross different run).

I prefere writing my lexical comparator more closely to the corresponding order:

``` c++
bool LineGroup::operator<(const LineGroup& other) const {
    if (this->name != other.name) { return this->name < other.name; }
    return this->idx < other.idx;
}
```

Commit:
ed: change database representation for line_groups

Keep consistency with NTFS representation.
Remove is_main_line in line_group_link and add main_line_id in line_group.
Remove is_main_line in ed types.
Fix fusio parser test and use english in fixture.

Commit:
jormungandr: rename groups to line_groups

Force depth of 0 for main_line in pb_converter to not duplicate too many informations.
Update commit of navitia-proto submodule.

eturck commented:
I think I have made all the changes : 
- Line_group table has a main_line_id column,
- Line_group_link doesn't have a is_main_line column anymore,
- Is_main_line is removed everywhere in the code since we don't use it anymore (neither in the database nor the json output),
- Using a flat_set instead of a vector for linked_lines_by_line_group_uri,
- Rename a few "groups" to "line_groups" (especially in the lines API),
- Force a depth of 0 for the main_line in line_groups API to reduce duplicated informations,
- Add 2 throw in NTFS file parsing when the header is not correct,
- Fix comparator to not compare pointer and compare idx instead,
- Translate line_groups.txt in ntfs fixture.

antoine-de commented:
Jenkins, test this please

Commit:
type: Update data version

antoine-de commented on source/ed/types.h:
```diff
@@ -224,6 +224,19 @@ struct Line : public Header, Nameable {
 
+struct LineGroup : public Header, Nameable {
+    const static nt::Type_e type = nt::Type_e::LineGroup;
+    std::string name;
```
you already have a name with the Nameable struct.

I'm not sure you want to inherit from this one, it's up to you.

antoine-de commented on source/jormungandr/tests/check_utils.py:
```diff
@@ -598,6 +598,14 @@ def is_valid_line(line, depth_check=1):
+    get_not_null(line_group, "id")
+
+    if depth_check > 0:
+        is_valid_line(get_not_null(line_group, "main_line"), depth_check - 1)
```
the main_line is with depth_check = 0 no ?

antoine-de commented on source/jormungandr/tests/ptref_tests.py:
```diff
@@ -119,6 +119,28 @@ def test_line(self):
         assert physical_modes[0]['id'] == 'physical_mode:Car'
         assert physical_modes[0]['name'] == 'name physical_mode:Car'
 
+        group = get_not_null(l, 'line_groups')
```
shouldn't you call is_valid_line_group there too?

antoine-de commented on source/jormungandr/tests/ptref_tests.py:
```diff
@@ -119,6 +119,28 @@ def test_line(self):
+        com = get_not_null(lg, 'comments')
+        assert len(com) == 1
+        assert com[0]['type'] == 'standard'
+        assert com[0]['value'] == "I'm a happy comment"
```
it's really great that you added some jormungandr tests! Since we decided to do some non conventional stuffs with the depth I think it would be great to add some more tests:
- one test per depth there just to be sure (after the call, a call to is_valid_line_group should be enough)
- some tests to check that the ptref stuff work well with the line_group: 
  - `self.query_region("v1/line_groups/{line group id}/lines")`
  - and maybe `self.query_region("v1/routes/{one route id}/line_groups")` to check if you can get all line_groups for a route

antoine-de commented on source/type/pb_converter.cpp:
```diff
@@ -409,6 +413,30 @@ void fill_pb_object(nt::Line const* l, const nt::Data& data,
+        return ;
+    int depth = (max_depth <= 3) ? max_depth : 3;
+
+    for (const auto& comment: data.pt_data->comments.get(lg)) {
```
comments for depth = 0 ? not sure it's really useful but you know better than me :smile:

antoine-de commented on source/type/type.h:
```diff
@@ -616,12 +620,15 @@ struct Line : public Header, Nameable, HasMessages, Codes{
 
     std::map<std::string,std::string> properties;
 
+    //The bool part of the pair stand for "is_main_line"
```
it seems there is no bool anymore :wink:

Commit:
jormungandr: add test on line_groups

Check if the main_line always have a depth of 0 when calling v1/lines_groups
Check if line_groups are accessible through the ptref graph
Remove old comment / code.

eturck commented:
New update!
- Add some tests recommanded by @antoine-de,
- Add a case in type.cpp in order to have line_groups accessible through the graph (and move a data->complete in main_ptref_test.cpp),
- Remove old / useless code / comment

antoine-de commented:
Nice, thanks @eturck !
